### PR TITLE
[scroll-animations-1] Add Element-based offset (#4337)

### PR DESCRIPTION
**Pull request by majido**
Add basic definition for Element-based offsets

Major changes:
* Introduce concept of "scroll timeline offset" that can be container-based (existing concept) and element-based (new concept).
* Add IDL for the new offset type and use it.
* Define the process for each offset type to be resolved into an effective scroll offset.
* Update current time calculation to resolve offsets and use the effective values.
* Add basic diagram to show the behavior visually for a simple example

Minor changes:
 * Rewrap lines to fit in 80 chars
 * Trim end-of-line whitespace
 * Clarify some definitions

TODO (as follow ups):
 - Define threshold for element-based offset.
 - Add css syntax for element-based offsets.
 - Add more examples

### CODE REVIEW

**Comment by majido** [comment]
@flackr  FYI in case you have any comments.

File: scroll-animations-1/Overview.bs [hunk 1]
```diff
@@ -391,69 +393,224 @@ not by wall-clock time, but by the progress of scrolling in a [=scroll container
...
+
+<div class=members>
+
+:   <dfn dict-member for=ElementBasedOffset>target</dfn>
+::  The target whose intersection with {{source}}'s [=scrolling box=] determines
+     the concrete scroll offset.
+
+:   <dfn dict-member for=ElementBasedOffset>edge</dfn>
+::  The edge of {{source}}'s [=scrolling box=] which the target should
+    intersection with.
```
Comment (ogerchikov) [inline scroll-animations-1/Overview.bs:512]: nit: intersect

File: scroll-animations-1/Overview.bs [hunk 2]
```diff
@@ -391,69 +393,224 @@ not by wall-clock time, but by the progress of scrolling in a [=scroll container
...
+    determined by {{orientation}}.
+
+1.  Clamp the value of <var>scroll offset</var> to be within the {{source}}'s
+     scroll range.
+
+1.  The [=effective scroll offset=] is <var>scroll offset</var>
+
+
+
+Note: The current algorithm selects the effective scorll offset such that the
```
Comment (ogerchikov) [inline scroll-animations-1/Overview.bs:560]: nit: scroll

File: scroll-animations-1/Overview.bs [hunk 3]
```diff
@@ -391,69 +393,224 @@ not by wall-clock time, but by the progress of scrolling in a [=scroll container
...
+  Usage of element-based offsets to create enter/exit triggers.<br>
+  The left figure shows the scroller and target being aligned at 'end' edge.<br>
+  The right figure shows them being aligned at 'start' edge.
+ </figcaption>
+</figure>
+
+
+Note that here we are expecting a typical top to bottom scrolling and thus
+consider the entrance to coincide when target's start edge is flushed with
+with scrollport's <strong>end edge</strong> and viceversa for exit.
```
Comment (ogerchikov) [inline scroll-animations-1/Overview.bs:583]: nit: remove extra "with"

**Comment by majido** [comment]
Addressed nits in latest commit https://github.com/w3c/csswg-drafts/pull/5124/commits/be7fcd7c8dc47158d307fbff1c5e381d7397d683

File: scroll-animations-1/Overview.bs [hunk 4]
```diff
@@ -344,8 +345,8 @@ enum ScrollTimelineAutoKeyword { "auto" };
 dictionary ScrollTimelineOptions {
   Element? source = null;
   ScrollDirection orientation = "block";
-  DOMString start = "auto";
-  DOMString end = "auto";
+  (DOMString or ElementBasedOffset) start = "auto";
+  (DOMString or ElementBasedOffset) end = "auto";
```
Comment (birtles) [inline scroll-animations-1/Overview.bs:349]: @majido Should this patch have updated the types of the `start` and `end` members of the `ScrollTimeline` interface too?

Comment (majido) [reply scroll-animations-1/Overview.bs:349]: Indeed! And thanks for the patch to fix it.

**Pull request merged**